File: src/FluidScript.Core/Layout/Scene.cs
using System.Collections.Immutable;

namespace FluidScript.Core.Layout;

/// <summary>An axis-aligned box in world units, <c>y</c> up (<c>28</c> §1).</summary>
/// <param name="X">The left edge.</param>
/// <param name="Y">The bottom edge.</param>
/// <param name="Width">The width; never negative.</param>
/// <param name="Height">The height; never negative.</param>
public readonly record struct Box(double X, double Y, double Width, double Height)
{
    /// <summary>Gets the right edge.</summary>
    public double Right => X + Width;

    /// <summary>Gets the top edge.</summary>
    public double Top => Y + Height;

    /// <summary>Gets the bottom edge, the same as <see cref="Y"/>.</summary>
    public double Bottom => Y;

    /// <summary>Gets the centre.</summary>
    public Point Centre => new(X + (Width / 2), Y + (Height / 2));

    /// <summary>A box of this size centred at a point.</summary>
    /// <param name="centre">The centre.</param>
    /// <param name="width">The width.</param>
    /// <param name="height">The height.</param>
    /// <returns>The box.</returns>
    public static Box Around(Point centre, double width, double height) =>
        new(centre.X - (width / 2), centre.Y - (height / 2), width, height);

    /// <summary>The smallest box holding every point.</summary>
    /// <param name="points">At least one point.</param>
    /// <returns>The bounds.</returns>
    public static Box Of(IEnumerable<Point> points)
    {
        var minX = double.PositiveInfinity;
        var minY = double.PositiveInfinity;
        var maxX = double.NegativeInfinity;
        var maxY = double.NegativeInfinity;

        foreach (var p in points)
        {
            minX = Math.Min(minX, p.X);
            minY = Math.Min(minY, p.Y);
            maxX = Math.Max(maxX, p.X);
            maxY = Math.Max(maxY, p.Y);
        }

        return double.IsInfinity(minX) ? new Box(0, 0, 0, 0) : new Box(minX, minY, maxX - minX, maxY - minY);
    }

    /// <summary>This box grown by <paramref name="margin"/> on every side.</summary>
    /// <param name="margin">The margin, world units.</param>
    /// <returns>The grown box.</returns>
    public Box Grow(double margin) => new(X - margin, Y - margin, Width + (2 * margin), Height + (2 * margin));

    /// <summary>This box moved.</summary>
    /// <param name="dx">East.</param>
    /// <param name="dy">North.</param>
    /// <returns>The moved box.</returns>
    public Box Offset(double dx, double dy) => new(X + dx, Y + dy, Width, Height);

    /// <summary>Whether two boxes share interior area; touching edges do not count.</summary>
    /// <param name="other">The other box.</param>
    /// <returns><see langword="true"/> when the interiors intersect.</returns>
    public bool Intersects(Box other) =>
        X < other.Right - 1e-9 && other.X < Right - 1e-9 && Y < other.Top - 1e-9 && other.Y < Top - 1e-9;

    /// <summary>The clearance between two boxes (<c>28</c> §23): the largest axis gap, negative when they overlap.</summary>
    /// <param name="other">The other box.</param>
    /// <returns>Two boxes are at least <c>g</c> apart exactly when this is at least <c>g</c>.</returns>
    public double GapTo(Box other) =>
        Math.Max(Math.Max(other.X - Right, X - other.Right), Math.Max(other.Y - Top, Y - other.Top));

    /// <summary>Whether a point lies strictly inside.</summary>
    /// <param name="point">The point.</param>
    /// <returns><see langword="true"/> when inside, not on the edge.</returns>
    public bool ContainsInterior(Point point) =>
        point.X > X + 1e-9 && point.X < Right - 1e-9 && point.Y > Y + 1e-9 && point.Y < Top - 1e-9;

    /// <summary>The smallest box holding both.</summary>
    /// <param name="other">The other box.</param>
    /// <returns>The union's bounds.</returns>
    public Box Union(Box other)
    {
        var x = Math.Min(X, other.X);
        var y = Math.Min(Y, other.Y);
        return new Box(x, y, Math.Max(Right, other.Right) - x, Math.Max(Top, other.Top) - y);
    }
}

/// <summary>A point in world units, <c>y</c> up.</summary>
/// <param name="X">Horizontal, growing east.</param>
/// <param name="Y">Vertical, growing north.</param>
public readonly record struct Point(double X, double Y)
{
    /// <summary>This point moved.</summary>
    /// <param name="dx">East.</param>
    /// <param name="dy">North.</param>
    /// <returns>The moved point.</returns>
    public Point Offset(double dx, double dy) => new(X + dx, Y + dy);

    /// <summary>This point moved along a direction.</summary>
    /// <param name="direction">The direction.</param>
    /// <param name="distance">World units; negative moves the other way.</param>
    /// <returns>The moved point.</returns>
    public Point Towards(Direction direction, double distance) => new(X + (direction.X * distance), Y + (direction.Y * distance));

    /// <summary>The Manhattan distance to another point.</summary>
    /// <param name="other">The other point.</param>
    /// <returns><c>|Δx| + |Δy|</c>.</returns>
    public double ManhattanTo(Point other) => Math.Abs(X - other.X) + Math.Abs(Y - other.Y);
}

/// <summary>Where a port meets its placed symbol, in world units, which way a pipe leaves it and which way the fluid flows (<c>28</c> §2, §5).</summary>
/// <param name="At">The inner anchor: the point on the inner box's edge.</param>
/// <param name="Direction">The outward unit vector, one of the four axis directions; the stub to the outer anchor runs along it.</param>
/// <param name="Flow">The flow vector: <see cref="Direction"/> for an outlet, its opposite for an inlet.</param>
public readonly record struct PlacedAnchor(Point At, Point Direction, Direction Flow)
{

    /// <summary>The point <paramref name="distance"/> along the outward direction; at the margin it is the outer anchor.</summary>
    /// <param name="distance">World units.</param>
    /// <returns>The point.</returns>
    public Point Along(double distance) => At.Offset(Direction.X * distance, Direction.Y * distance);

    /// <summary>Gets the outward direction as a <see cref="Layout.Direction"/>.</summary>
    public Direction Outward => Layout.Direction.Of(Direction) ?? Layout.Direction.Right;
}

/// <summary>One component's place in the diagram (<c>D-103</c>).</summary>
/// <remarks>
/// <see cref="Inner"/> is the symbol's box scaled, rotated and given an arrangement; <see cref="Outer"/> is
/// the inner grown by the margin. The invariant the solver holds (<c>28</c> §4): every two inner boxes are at
/// least the clearance apart; outer boxes may overlap, and that is counted, not rejected.
/// </remarks>
public sealed record Placement
{
    /// <summary>The component's name.</summary>
    public required string ComponentId { get; init; }

    /// <summary>The symbol definition drawn inside <see cref="Inner"/>.</summary>
    public required string SymbolId { get; init; }

    /// <summary>The symbol's box after placement.</summary>
    public required Box Inner { get; init; }

    /// <summary>The inner box grown by the margin: the clearance nothing else may enter.</summary>
    public required Box Outer { get; init; }

    /// <summary>The quarter turn applied, clockwise degrees: 0, 90, 180 or 270.</summary>
    public required int Rotation { get; init; }

    /// <summary>Whether the symbol was mirrored left-to-right before rotating.</summary>
    public required bool Mirrored { get; init; }

    /// <summary><c>default</c> or one of the symbol's alternative arrangements (<c>D-102</c>).</summary>
    public required string Arrangement { get; init; }

    /// <summary>Every port's anchor after placement, by port name; a node's ports each get their own entry.</summary>
    public required ImmutableSortedDictionary<string, PlacedAnchor> Anchors { get; init; }

    /// <summary>Where the label sits.</summary>
    public required Point LabelAt { get; init; }

    /// <summary><c>computed</c>; <c>pinned</c> is reserved for a placement the script states (<c>D-103</c>).</summary>
    public required string Source { get; init; }


    /// <summary>The <see cref="LayoutGroup.Id"/> of the group that placed this component, or <see langword="null"/> when it was placed on its own.</summary>
    public string? Group { get; init; }

    /// <summary>Whether this is an inline element (<c>D-105</c>): a pipe or a two-port node drawn as a point on its run, with no box and no clearance of its own.</summary>
    public bool IsInline => Inner.Width <= 0 && Inner.Height <= 0;
}

/// <summary>One connection's path, in world units.</summary>
/// <param name="ConnectionId"><c>c{n}</c>, or a non-flow element's id for a signal line.</param>
/// <param name="Kind"><c>pipe</c> for a flow connection, <c>signal</c> for an instrument's leader line.</param>
/// <param name="Points">The polyline, orthogonal segment by segment and normalised (<c>28</c> §20); the first and last points are the inner anchors.</param>
/// <param name="Layer">The draw order (<c>28</c> C16): <c>supply</c> in front, <c>return</c> behind it, <c>signal</c> behind everything.</param>
/// <param name="Hops">Where this route passes behind another it crosses; the picture breaks this route around each.</param>
public sealed record Route(string ConnectionId, string Kind, string Layer, ImmutableArray<Point> Points, ImmutableArray<Point> Hops)
{
    /// <summary>Gets the total length, world units.</summary>
    public double Length
    {
        get
        {
            var length = 0.0;
            for (var i = 1; i < Points.Length; i++)
            {
                length += Points[i - 1].ManhattanTo(Points[i]);
            }

            return length;
        }
    }

    /// <summary>Gets the interior points where the direction changes.</summary>
    public IEnumerable<Point> Bends
    {
        get
        {
            for (var i = 1; i + 1 < Points.Length; i++)
            {
                var a = Points[i - 1];
                var b = Points[i];
                var c = Points[i + 1];
                var horizontalIn = Math.Abs(b.Y - a.Y) < 1e-9;
                var horizontalOut = Math.Abs(c.Y - b.Y) < 1e-9;

                if (horizontalIn != horizontalOut)
                {
                    yield return b;
                }
            }
        }
    }
}

/// <summary>A group the solver laid out as one object (<c>28</c> §24–25): its members were placed together and the parent only moved them.</summary>
/// <param name="Id">A stable name: <c>loop-1</c>, <c>header-1</c>, <c>chain-3</c>.</param>
/// <param name="Kind"><c>loop</c>, <c>header</c>, <c>chain</c> or <c>fragment</c>.</param>
/// <param name="Orientation"><c>cw</c> or <c>ccw</c> for a loop; the flow direction for a chain; empty otherwise.</param>
/// <param name="Members">The component ids, in the group's own order.</param>
/// <param name="Bounds">The union of the members' inner boxes and the group's own routes.</param>
public sealed record LayoutGroup(string Id, string Kind, string Orientation, ImmutableArray<string> Members, Box Bounds);

/// <summary>The whole drawing: what the renderer draws and the exporter writes (<c>D-103</c>).</summary>
public sealed record Scene
{
    /// <summary>Every component's placement, in <c>hints.Order</c> then non-flow elements.</summary>
    public required ImmutableArray<Placement> Placements { get; init; }

    /// <summary>Every connection's route, in connection order, then signal lines.</summary>
    public required ImmutableArray<Route> Routes { get; init; }

    /// <summary>The bounds of everything, outer boxes and routes included.</summary>
    public required Box Extent { get; init; }

    /// <summary>The margin every outer box was grown by, world units.</summary>
    public required double Margin { get; init; }

    /// <summary>The groups the solver laid out as objects, outermost first; empty for a scene with none.</summary>
    public ImmutableArray<LayoutGroup> Groups { get; init; } = [];
}

File: frontend/src/features/canvas/scene.ts
import type {
  Component,
  Diagnostic,
  ModelContract,
  Placement,
  Primitive,
  ResolvedStyle,
  Route,
  Scale,
  ScalePosition,
  SymbolDefinition,
} from '../../api/types.ts';
import {
  boxOf,
  centreOf,
  grow,
  isDegenerate,
  pointOf,
  pointsOf,
  type Box,
  type Point,
} from './geometry.ts';

/**
 * The prepared scene (`53`, `D-71`): what the canvas draws, the exporter serializes and the golden
 * test pins. Every coordinate in it is the layout's (`D-103`, `53` invariant 2): the renderer
 * chooses nothing about where things are, only how the marks look. Pure of the viewport and of
 * React, so a headless test and the export share it (`53` invariant 5).
 */
export interface PreparedScene {
  /**
   * Whether the model behind the drawing solved: `false` when nothing was solved (the script was
   * refused before the solver) or the solver did not converge. An unsolved plant is drawn in the
   * error colour, so the reader sees it without opening the log (the user's ask, 2026-09-18).
   */
  readonly solved: boolean;
  /** The layout's extent grown by the margin, world units, y up. */
  readonly bounds: Box;
  readonly margin: number;
  /** Every placement with a box, in the layout's order. */
  readonly symbols: readonly PreparedSymbol[];
  /** Inline placements (`D-105`): a boundary node's dot, a pipe's label point. */
  readonly marks: readonly PreparedMark[];
  /** Every route, from the back: signals, return, supply (`28` C16). */
  readonly routes: readonly PreparedRoute[];
  readonly labels: readonly PreparedLabel[];
  /** The property the colours follow: the script's `show`, or the reader's switch (`57`, `D-117`). */
  readonly property: string;
  /** The scale behind the colours, for the legend; `null` when the wire has none for the property. */
  readonly scale: Scale | null;
  /** The properties the legend's switcher offers. */
  readonly available: readonly string[];
}

export type Severity = 'error' | 'warning';

export interface PreparedSymbol {
  /** The component id, the only key (`D-34`). */
  readonly id: string;
  readonly kind: string;
  readonly symbolId: string;
  /** The strokes, or `null` for a kind the wire's symbol table does not hold (`53` error cases). */
  readonly primitives: readonly Primitive[] | null;
  readonly inner: Box;
  readonly centre: Point;
  /** Clockwise quarter turns as the layout states them. */
  readonly rotation: number;
  readonly mirrored: boolean;
  readonly inferred: boolean;
  readonly junction: boolean;
  /** The scale's position of the representative value, 0 to 1, for the `state` fill slot (`57`); `null` when unsolved. */
  readonly scale: number | null;
  /** The inlet's and the outlet's positions where the component has both (an exchanger): the fill is a gradient from one to the other (`57` Components). */
  readonly scaleFrom: number | null;
  readonly scaleTo: number | null;
  /** The worst diagnostic addressed to the component (`R-24`). */
  readonly badge: Severity | null;
  /** True when a parameter of the component was sized or defaulted rather than stated (`D-02`). */
  readonly sized: boolean;
  readonly ports: readonly PreparedPort[];
  readonly style: ResolvedStyle | null;
}

export interface PreparedPort {
  readonly name: string;
  readonly at: Point;
  readonly direction: Point | null;
}

export interface PreparedMark {
  readonly id: string;
  readonly kind: 'node' | 'pipe';
  readonly at: Point;
  readonly inferred: boolean;
  /** A node with one connection is a boundary and draws a dot; with two it draws nothing (`D-105`). */
  readonly boundary: boolean;
}

export interface PreparedRoute {
  readonly id: string;
  readonly kind: 'pipe' | 'signal';
  readonly layer: string;
  /** The polyline cut around the hops this route owns; each piece is drawn on its own. */
  readonly pieces: readonly (readonly Point[])[];
  readonly arrow: 'forward' | 'reverse' | 'none';
  readonly corner: 'fillet' | 'round' | 'sharp';
  readonly scaleFrom: number | null;
  readonly scaleTo: number | null;
  readonly style: ResolvedStyle | null;
}

export interface PreparedLabel {
  readonly ownerId: string;
  /** The tag where the component has one, else its id (`D-34`); display only. */
  readonly text: string;
  /** The centre of `box`. */
  readonly at: Point;
  /** The box the layout reserved for the text from the declared metric (`D-73`, `C-84`). */
  readonly box: Box;
  /** False when the layout could not place the label clear; the canvas then draws a leader to `owner`. */
  readonly clear: boolean;
  /** The owner's box, where a leader points. */
  readonly owner: Point;
  readonly inferred: boolean;
  /** True for an inline element's label, shown only above `53`'s 3× level of detail. */
  readonly inline: boolean;
}

const layerOrder: Readonly<Record<string, number>> = { supply: 2, return: 1 };

/**
 * Prepares the scene of a model. A model with `solved: false` prepares the same way: topology with
 * no state (`53` invariant 4). `property` is the reader's switch (`57` invariant 6): every available
 * scale is on the wire (`D-117`), so choosing another is a re-preparation, never a request; a
 * property the wire does not carry falls back to the script's.
 */
export function prepareScene(model: ModelContract, property?: string): PreparedScene {
  const layout = model.layout;
  const margin = layout.margin;
  const visualization = model.visualization;
  const shown =
    property !== undefined && property in visualization.scales ? property : visualization.active;
  const positionOf = (scales: Placement['scales'] | Route['scales']): ScalePosition =>
    scales[shown] ?? { at: null, from: null, to: null };
  const components = new Map<string, Component>(model.components.map((c) => [c.id, c]));
  const symbols = new Map<string, SymbolDefinition>(model.symbols.map((s) => [s.id, s]));
  const badges = worstBySubject(model.diagnostics);
  const connectionsOf = countConnections(model);

  const preparedSymbols: PreparedSymbol[] = [];
  const marks: PreparedMark[] = [];
  const labels: PreparedLabel[] = [];

  for (const placement of layout.placements) {
    const component = components.get(placement.componentId);
    const inferred = component?.origin.startsWith('inferred') ?? true;
    const inner = boxOf(placement.inner);
    const labelAt = pointOf(placement.labelAt);
    const labelBox = boxOf(placement.labelBox);
    const labelClear = placement.labelClear;
    const text = component?.tag ?? placement.componentId;

    if (isDegenerate(inner)) {
      const kind = placement.symbolId.startsWith('pipe') ? 'pipe' : 'node';
      marks.push({
        id: placement.componentId,
        kind,
        at: centreOf(inner),
        inferred,
        boundary: kind === 'node' && (connectionsOf.get(placement.componentId) ?? 0) < 2,
      });
      labels.push({
        ownerId: placement.componentId,
        text,
        at: labelAt,
        box: labelBox,
        clear: labelClear,
        owner: centreOf(inner),
        inferred,
        inline: true,
      });
      continue;
    }

    const definition = symbols.get(placement.symbolId);
    preparedSymbols.push({
      id: placement.componentId,
      kind: component?.kind ?? placement.symbolId,
      symbolId: placement.symbolId,
      primitives: definition?.primitives ?? null,
      inner,
      centre: centreOf(inner),
      rotation: placement.rotation,
      mirrored: placement.mirrored,
      inferred,
      junction: Object.keys(placement.anchors).length >= 3 && placement.symbolId.startsWith('node'),
      scale: positionOf(placement.scales).at,
      scaleFrom: positionOf(placement.scales).from,
      scaleTo: positionOf(placement.scales).to,
      badge: badges.get(placement.componentId) ?? null,
      sized: component !== undefined && isSized(component),
      ports: portsOf(placement),
      style: placement.style ?? null,
    });
    labels.push({
      ownerId: placement.componentId,
      text,
      at: labelAt,
      box: labelBox,
      clear: labelClear,
      owner: centreOf(inner),
      inferred,
      inline: false,
    });
  }

  const corner = cornerOf(model.style.default.corner) ?? 'sharp';
  const inline = new Set(marks.map((m) => m.id));
  const arrowed = routesThatCarryTheArrow(model, inline);
  const routes = [...layout.routes]
    .sort((a, b) => (layerOrder[a.layer] ?? 0) - (layerOrder[b.layer] ?? 0))
    .map((route) =>
      prepareRoute(
        route,
        margin / 4,
        arrowed.has(route.id) ? layout.flow[route.id] : undefined,
        cornerOf(route.style?.corner) ?? corner,
        positionOf(route.scales),
      ),
    );

  return {
    solved: model.solve?.converged === true,
    bounds: grow(boxOf(layout.extent), margin),
    margin,
    symbols: preparedSymbols,
    marks,
    routes,
    labels,
    property: shown,
    scale: visualization.scales[shown] ?? null,
    available: visualization.available,
  };
}

function portsOf(placement: Placement): PreparedPort[] {
  return Object.entries(placement.anchors).flatMap(([name, anchor]) =>
    anchor === undefined
      ? []
      : [
          {
            name,
            at: pointOf(anchor.at),
            direction:
              anchor.direction === null || anchor.direction === undefined
                ? null
                : pointOf(anchor.direction),
          },
        ],
  );
}

function prepareRoute(
  route: Route,
  gap: number,
  flow: string | undefined,
  corner: PreparedRoute['corner'],
  position: ScalePosition,
): PreparedRoute {
  return {
    id: route.id,
    kind: route.kind === 'signal' ? 'signal' : 'pipe',
    layer: route.layer,
    pieces: cutAtHops(pointsOf(route.points), pointsOf(route.hops), gap),
    arrow: route.kind === 'signal' ? 'none' : arrowOf(flow),
    corner,
    scaleFrom: position.from,
    scaleTo: position.to,
    style: route.style ?? null,
  };
}

/**
 * One arrow per drawn run, not per connection. The compiler splits a pipe the reader sees as one
 * line into several connections through inferred inline nodes and pipes (`D-105`, rule I7), each a
 * route with its own state; the arrow belongs to the run, so it goes on the longest route of the
 * chain between two drawn symbols and the others carry none.
 */
function routesThatCarryTheArrow(model: ModelContract, inline: ReadonlySet<string>): Set<string> {
  const routeOf = new Map(model.layout.routes.map((r) => [r.id, r]));
  const connectionOf = new Map(model.connections.map((c) => [c.id, c]));
  const byInlineEnd = new Map<string, string[]>();
  for (const connection of model.connections) {
    for (const end of [connection.from.component, connection.to.component]) {
      if (inline.has(end)) {
        byInlineEnd.set(end, [...(byInlineEnd.get(end) ?? []), connection.id]);
      }
    }
  }
  const length = (id: string): number => {
    const points = pointsOf(routeOf.get(id)?.points ?? []);
    let total = 0;
    for (let i = 1; i < points.length; i++) {
      total +=
        Math.abs(points[i]!.x - points[i - 1]!.x) + Math.abs(points[i]!.y - points[i - 1]!.y);
    }
    return total;
  };

  const seen = new Set<string>();
  const carriers = new Set<string>();
  for (const connection of model.connections) {
    if (seen.has(connection.id)) {
      continue;
    }
    // The run: every connection reachable through inline ends, in both directions.
    const chain = new Set<string>([connection.id]);
    const queue = [connection.id];
    while (queue.length > 0) {
      const c = connectionOf.get(queue.pop()!)!;
      for (const end of [c.from.component, c.to.component]) {
        for (const next of byInlineEnd.get(end) ?? []) {
          if (!chain.has(next)) {
            chain.add(next);
            queue.push(next);
          }
        }
      }
    }
    let longest: string | null = null;
    for (const id of chain) {
      seen.add(id);
      if (longest === null || length(id) > length(longest)) {
        longest = id;
      }
    }
    if (longest !== null) {
      carriers.add(longest);
    }
  }
  return carriers;
}

function arrowOf(flow: string | undefined): PreparedRoute['arrow'] {
  return flow === 'forward' || flow === 'reverse' ? flow : 'none';
}

function cornerOf(corner: string | null | undefined): PreparedRoute['corner'] | null {
  return corner === 'fillet' || corner === 'round' || corner === 'sharp' ? corner : null;
}

/**
 * Breaks a route around each hop it owns, a quarter margin either side, so the route in front runs
 * through the gap (`28` C16). A port of the Core instrument's `Pieces`, so both pictures agree.
 */
export function cutAtHops(
  points: readonly Point[],
  hops: readonly Point[],
  gap: number,
): Point[][] {
  const first = points[0];
  if (first === undefined) {
    return [];
  }
  const pieces: Point[][] = [];
  let piece: Point[] = [first];

  for (let i = 1; i < points.length; i++) {
    const a = points[i - 1]!;
    const b = points[i]!;
    const dx = Math.sign(b.x - a.x);
    const dy = Math.sign(b.y - a.y);
    const along = (p: Point): number => (p.x - a.x) * dx + (p.y - a.y) * dy;
    const onSegment = hops
      .filter(
        (h) =>
          Math.abs((h.x - a.x) * dy - (h.y - a.y) * dx) < 1e-9 &&
          along(h) > 0 &&
          (b.x - h.x) * dx + (b.y - h.y) * dy > 0,
      )
      .sort((p, q) => along(p) - along(q));

    for (const h of onSegment) {
      piece.push({ x: h.x - dx * gap, y: h.y - dy * gap });
      pieces.push(piece);
      piece = [{ x: h.x + dx * gap, y: h.y + dy * gap }];
    }
    piece.push(b);
  }

  pieces.push(piece);
  return pieces;
}

function worstBySubject(diagnostics: readonly Diagnostic[]): Map<string, Severity> {
  const out = new Map<string, Severity>();
  for (const d of diagnostics) {
    const subject = d.component;
    if (subject === null || subject === undefined || d.severity === 'info') {
      continue;
    }
    if (d.severity === 'error' || !out.has(subject)) {
      out.set(subject, d.severity === 'error' ? 'error' : 'warning');
    }
  }
  return out;
}

function countConnections(model: ModelContract): Map<string, number> {
  const out = new Map<string, number>();
  for (const connection of model.connections) {
    for (const end of [connection.from, connection.to]) {
      out.set(end.component, (out.get(end.component) ?? 0) + 1);
    }
  }
  return out;
}

function isSized(component: Component): boolean {
  return Object.values(component.parameters).some((p) => p !== undefined && p.source !== 'stated');
}
